python/paddle/fluid/backward.py

```diff
@@ -1101,67 +1102,119 @@ def append_backward(loss,
         isinstance(callbacks, list)
 
     program = loss.block.program
+
+    # Todo(liym27): when to add 1 in control flow
     program._appending_grad_times += 1
 
     if no_grad_set is None:
         no_grad_set = set()
     no_grad_set = copy.copy(no_grad_set)
     no_grad_dict = _get_stop_gradients_(program)
+    # no_grad_set only contain vars in block 0
+    # Todo(liym27): support vars in sub block
     no_grad_dict[0].update(list(map(_append_grad_suffix_, no_grad_set)))
 
     root_block = program.block(0)
 
-    fwd_op_num = root_block.desc.op_size()
     current_block_idx = program.current_block_idx
+
+    current_block = program.block(current_block_idx)
+
+    # create grad block if in control flow
```

Comment:
Please add comment about the use case: currently it is only to support the optimizer.minimize in a switch branch, which can append_backward in a sub_block.

We had better to explain it because while_loop is also in control flow, but adding optimizer in while makes no sense. We should report error when it is in while_loop. I will ask Fei to do it. You can just fully explain it or add a TODO about while_loop here.

Reply:
Thanks. Done